protocol/messenger_communities.go

```diff
@@ -2437,129 +2414,178 @@ func (m *Messenger) RequestImportDiscordCommunity(request *requests.ImportDiscor
-		totalCategoriesCount := len(exportData.Categories)
 
-		for _, category := range exportData.Categories {
+		for i, importFile := range request.FilesToImport {
 
-			createCommunityCategoryRequest := &requests.CreateCommunityCategory{
-				CommunityID:  discordCommunity.ID(),
-				CategoryName: category.Name,
-				ChatIDs:      make([]string, 0),
-			}
-			// We call `CreateCategory` on `communitiesManager` directly so we can control
-			// whether or not the community update should be published (it should not until the
-			// import has finished)
-			communityWithCategories, changes, err := m.communitiesManager.CreateCategory(createCommunityCategoryRequest, false)
-			if err != nil {
-				m.cleanUpImport(communityID)
-				importProgress.AddTaskError(discord.CommunityCreationTask, discord.Error(err.Error()))
-				importProgress.StopTask(discord.CommunityCreationTask)
+			exportData, errs := m.ExtractDiscordDataFromImportFiles([]string{importFile})
+			if len(errs) > 0 {
+				for _, err := range errs {
+					importProgress.AddTaskError(discord.CommunityCreationTask, err)
+				}
 				progressUpdates <- importProgress
 				return
 			}
-			discordCommunity = communityWithCategories
-			// This looks like we keep overriding the same field but there's
-			// only one `CategoriesAdded` change at this point.
-			for _, addedCategory := range changes.CategoriesAdded {
-				processedCategoriesIds[category.ID] = addedCategory.CategoryId
+			totalChannelsCount := len(exportData.ExportedData)
+			totalMessageCount := exportData.MessageCount
+
+			if totalChannelsCount == 0 || totalMessageCount == 0 {
+				importError := discord.Error(fmt.Errorf("No channel to import messages from in file: %s", importFile).Error())
+				if totalMessageCount == 0 {
+					importError.Message = fmt.Errorf("No messages to import in file: %s", importFile).Error()
+				}
+				importProgress.AddTaskError(discord.ChannelsCreationTask, importError)
+				progressUpdates <- importProgress
+				continue
 			}
 
-			// We're multiplying `progressValue` by 0.25 as it's added to the previous 0.75 progress
-			progressValue := (float32(len(processedCategoriesIds)) / float32(totalCategoriesCount)) * 0.25
-			importProgress.UpdateTaskProgress(discord.CommunityCreationTask, 0.75+progressValue)
+			importProgress.CurrentChunk = i + 1
 
-			progressUpdates <- importProgress
-		}
+			// We actually only ever receive a single category
+			// from `exportData` but since it's a map, we still have to
+			// iterate over it to access its values
+			for _, category := range exportData.Categories {
 
-		if m.DiscordImportMarkedAsCancelled(communityID) {
-			importProgress.StopTask(discord.CommunityCreationTask)
-			progressUpdates <- importProgress
-			cancel <- communityID
-			return
-		}
+				categories := discordCommunity.Categories()
+				exists := false
+				for catID := range categories {
+					if strings.HasSuffix(catID, category.ID) {
+						exists = true
+						break
+					}
+				}
```

Comment:
@cammellos here's how we check if a given category was already created during the import